src/Storages/MergeTree/MergedPartOffsets.h

```diff
@@ -0,0 +1,271 @@
+        current_page_values.push_back(val);
+    }
+
+    /// Compresses and finalizes the current page of values.
+    /// Called automatically when a page is full or at the end to finalize the structure.
+    void flush()
+    {
+        if (current_page_values.empty())
+            return;
+
+        size_t bits_per_val = 64 - getLeadingZeroBits(current_page_values.back() - current_page_values.front());
+        chassert(bits_per_val >= 1);
+        size_t num_uint64 = ((current_page_values.size() * bits_per_val) + 63) / 64;
+        chassert(num_uint64 >= 1);
+
+        pages.emplace_back(
+            current_page_values,
+            bits_per_val,
+            reinterpret_cast<UInt64 *>(arena.alignedAlloc(num_uint64 * sizeof(UInt64), alignof(UInt64))));
+        current_page_values.clear();
+    }
+
+    /// Decompresses the _part_offset value at the specified index
+    UInt64 operator[](size_t i) const
+    {
+        size_t page_pos = i >> PACKED_PAGE_SIZE_DEGREE;
+        chassert(page_pos < pages.size());
+        size_t page_idx = i & PACKED_PAGE_MASK;
+        return pages[page_pos][page_idx];
+    }
+
+    void clearTemporaryStorage() { current_page_values = {}; }
+
+    size_t totalAllocatedMemory() const { return pages.allocated_bytes() + current_page_values.allocated_bytes() + arena.allocatedBytes(); }
+};
+
+
+/// Manages _part_offset mapping during data part merges.
+/// Tracks how rows from original parts are positioned in the merged result.
+/// Provides efficient lookup from original _part_offset to new _part_offset in merged data.
+class MergedPartOffsets
+{
+public:
+    MergedPartOffsets() = default;
+
+    /// @starting_offsets: Starting cumulative _part_offset for each data part.
+    /// @total_rows: Total number of rows across all merging parts.
+    MergedPartOffsets(std::vector<UInt64> starting_offsets_, size_t total_rows_)
+        : starting_offsets(std::move(starting_offsets_))
+        , total_rows(total_rows_)
+        , num_parts(starting_offsets.size())
+        , offset_maps(num_parts)
+    {
+    }
+
+    MergedPartOffsets(MergedPartOffsets && other) noexcept { std::swap(*this, other); }
+
+    MergedPartOffsets & operator=(MergedPartOffsets && other) noexcept
+    {
+        std::swap(*this, other);
+        return *this;
+    }
+
+    /// Records _part_offset mappings for a batch of cumulated _part_offset values.
+    /// Determines which part each offset belongs to and maintains mapping to new _part_offset.
+    void insert(const UInt64 * begin_cumulated_offset, const UInt64 * end_cumulated_offset)
+    {
+        for (const UInt64 * it = begin_cumulated_offset; it != end_cumulated_offset; ++it)
+        {
+            /// Find which original part this cumulated offset belongs to
+            size_t part_idx = 0;
+            while (part_idx < num_parts - 1)
+            {
+                if (starting_offsets[part_idx + 1] > *it)
```

Comment:
I was thinking about how to avoid iterating in a loop here.
One option is to encode part number into source part offsets somehow (e.g. for the part with the number `i` out of `N` we can encode as `value * N + i`).

But maybe, actually, we can use two virtual columns: `_part_num` and `_part_offset`? It seems to be relatively cheap, as (if I understand correctly) `_part_num` would exist only in currently processing chunks, and the only overhead is an additional shuffling (for sorting).

Reply:
I think it's possible to reuse `_part_index`. Will try to optimize and see how it goes.

Reply:
We can do it in the next PR. Not a blocker for merging.